src/protocols/rdp/channels/rdpsnd/rdpsnd.c

```diff
@@ -48,6 +52,10 @@ void guac_rdpsnd_process_receive(guac_rdp_common_svc* svc,
         guac_rdpsnd_wave_handler(svc, input_stream, &header);
         return;
     }
+    
+    /* Check body size */
+    if (Stream_GetRemainingLength(input_stream) < header.body_size)
```

Comment:
RDPSND is a bit convoluted in the way it handles each segment of sound data. I think the reason you're seeing this fail is that audio data is split across two PDUs:

1. "WaveInfo" (with type `SNDC_WAVE`).
2. "Wave" (the PDU immediately following a "WaveInfo" PDU)

In the case of "WaveInfo", the body size is not the size of that PDU, but the size of that PDU _plus_ the next:

> The BodySize field of the RDPSND PDU Header is the size of the WaveInfo PDU plus the size of the data field of the Wave PDU that immediately follows this packet minus the size of the Header.

Reply:
Given the above, my thinking would be that it's not correct to check against `header.body_size` here, but that each logical attempt to read/seek through a block of data within `input_stream` should take it upon itself to first verify that the necessary number of bytes are available before attempting that read/seek.

Reply:
Okay - moving it down to where I currently have it _seems_ to work, but if there's something different I should do, here, I'm happy to change it up.

Reply:
Well, the [documentation for RDPSND regarding the WaveInfo PDU](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpea/c53cd81c-0d7f-4e68-8b95-1c1da68dbaac
) pretty specifically states that the body size does not necessarily indicate the size of the remainder of the PDU:

> **Header (4 bytes):** An [RDPSND PDU Header (section 2.2.1)](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpea/863d154b-dbe0-4781-979d-fa48daed721e). The **msgType** field of the RDPSND PDU Header MUST be set to SNDC_WAVE (0x02). The **BodySize** field of the RDPSND PDU Header is the size of the WaveInfo PDU plus the size of the data field of the [Wave PDU](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpea/81841793-f5d3-4305-aa22-ddcbd81a96b5) that immediately follows this packet minus the size of the Header.

Given this, I don't believe it is correct to check `body_size` in this way, and I'm not sure why it would be working. The above indicates to me that `body_size` would always be larger than `Stream_GetRemainingLength()` for a WaveInfo PDU .

Reply:
So, sounds like I should pull it out of there?  Particularly since the Streams get checked elsewhere?  And it doesn't seem like it's possible to accurately check this size given that we have to also check the next packet?  Or should it be checked against some other value than `body_size`?

Reply:
Yeah, I think so. The value of `body_size` is needed, but as it doesn't (necessarily) correspond to the size of the PDU, it can't really be used as a sanity check for PDU size.

I think the critical thing is simply that a check of remaining buffer size is made before any attempt to read/seek through a block of the input stream. In the case of the code above, this would be both within that function (making sure there is enough space for the header that the function reads) and within the PDU-specific handlers invoked by that function (making sure there is enough space for whatever data those handlers read, which only those handlers should be aware of).